Decoupling from-font resolution from FontCascadeFonts::primaryFont

vitorroriz opened the pull request:
#### f923c324da2f95fbd856ac79fc43aeb662f172cc
<pre>
Decoupling from-font resolution from FontCascadeFonts::primaryFont
<a href="https://bugs.webkit.org/show_bug.cgi?id=273339">https://bugs.webkit.org/show_bug.cgi?id=273339</a>
<a href="https://rdar.apple.com/127132471">rdar://127132471</a>

Reviewed by Brent Fulgham.

font-size-adjust from-font resolution was fixed by [1]. It now works correctly and
it uses the primary font for resolution. However, this patch changed FontCascadeFonts::primaryFont
to receive a mutable (non-const) FontCascadeDescription object, such that it could resolve the
from-font reference after the primary font is resolved.

This means that now FontCascadeFonts::primaryFont does two different things:
a) resolving and caching a font as the primary font
b) resolving from-font according to the primary font if necessary

The (a) part of primaryFont doesn&apos;t require a mutable FontCascadeDescription.
We can update FontCascadeDescription to have a method for resolving font-size-adjust
from-font at the point where we own a mutable FontCascadeDescription instead.

This allows us to compute the primaryFont itself from places where we only have access
to a const FontCascadeDescription. This is for example, required for solving
<a href="https://bugs.webkit.org/show_bug.cgi?id=273233">https://bugs.webkit.org/show_bug.cgi?id=273233</a>, where we need to access to primaryFont
from FontCascadeFonts::glyphDataForVariant

This patch should have no side effects, as it is a preparation
for solving 273233.

[1] <a href="https://commits.webkit.org/269041@main">https://commits.webkit.org/269041@main</a>

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSizeAdjustFromStyle):
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::primaryFont const):
* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
(WebCore::FontCascadeDescription::resolveFontSizeAdjustFromFontIfNeeded):
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/platform/graphics/FontCascadeFonts.h:
(WebCore::FontCascadeFonts::primaryFont):
* Source/WebCore/platform/graphics/FontSizeAdjust.h:
(WebCore::FontSizeAdjust::resolveFromFontIfNeeded):
(WebCore::FontSizeAdjust::shouldResolveFromFont const):

Canonical link: <a href="https://commits.webkit.org/278130@main">https://commits.webkit.org/278130@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f53e50d179cb7f39f873cdb8ac376cd85f864072

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49596 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28881 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52643 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52838 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/272 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/51900 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34901 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26501 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40485 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51696 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26425 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/42722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/21600 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/23869 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/43908 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7967 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45786 "Build is in progress. Recent messages:") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/44414 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54415 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24681 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/20856 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/25950 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42861 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46881 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10884 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/26796 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/25675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/platform/graphics/FontCascade.h:
```diff
@@ -373,7 +373,9 @@ class FontCascade final : public CanMakeWeakPtr<FontCascade>, public CanMakeChec
 {
     ASSERT(m_fonts);
-    return protectedFonts()->primaryFont(m_fontDescription);
+    const auto& font = protectedFonts()->primaryFont(m_fontDescription);
```
We normally omit the `const`.

brentfulgham commented on Source/WebCore/platform/graphics/FontSizeAdjust.h:
```diff
@@ -44,7 +44,7 @@ struct FontSizeAdjust {
     };
 
-    std::optional<float> resolve(float computedSize, const FontMetrics& fontMetrics) const
+    std::optional<float> computeValue(float computedSize, const FontMetrics& fontMetrics) const
```
I think the original name `resolve` seems right, especially when it is used in a method called `resolveFromFontIfNeeded`.

brentfulgham commented on Source/WebCore/platform/graphics/FontSizeAdjust.h:
```diff
@@ -69,12 +69,23 @@ struct FontSizeAdjust {
     Metric metric { Metric::ExHeight };
     ValueType type { ValueType::Number };
     Markable<float, WTF::FloatMarkableTraits> value { };
+    bool hasFromFontResolved { false };
```
We do not need this `bool` to indicate the font has been resolved -- `Markable` is a flavor of `std::optional`, so we just need to check it as a `bool`, or `value_or`, or whatever.

vitorroriz replied:
This is a good idea, the only issue is that in the case that the value can't be resolved due to font-size-adjust not having a metric, we would always try to resolve it again.

That said, we always assigned to metricValue before (default is xHeight() when metric is not specified) so I think we can still apply your suggestion.

brentfulgham replied:
Ah! I see. So we can't tell the difference between the font being unresolved, and the font not having a metric? So the case would be a 'FromFont' type that doesn't have a relevant metric? Is that an allowable state?

vitorroriz replied:
I've edit my comment, I meant when font-size-adjust has no declared metric. However, looking again at specification, this is an optional parameter for the property, which is default to xHeight, so your suggestion is correct indeed.

brentfulgham commented on Source/WebCore/platform/graphics/FontSizeAdjust.h:
```diff
@@ -69,12 +69,23 @@ struct FontSizeAdjust {
+
     bool isNone() const { return !value && type != ValueType::FromFont; }
     bool isFromFont() const { return type == ValueType::FromFont; }
+    bool shouldResolveFromFont() const { return isFromFont() && !hasFromFontResolved; }
```
So this should just be:
```
{ return isFromFont() && !value; }
```

brentfulgham commented on Source/WebCore/platform/graphics/FontSizeAdjust.h:
```diff
@@ -69,12 +69,23 @@ struct FontSizeAdjust {
+        if (!shouldResolveFromFont())
+            return false;
+        value = computeValue(computedSize, fontMetrics);
+        hasFromFontResolved = true;
```
Once you assign `value` you don't need the separate `bool`. That's the whole point of `std::optional` (and therefore `Markable`).

brentfulgham reviewed (changes requested):
I'd like you to restore the original `resolve` name, and avoid adding the `bool` to `FontSizeAdjust`, since it already stores the `value` in the equivalent of a `std::optional`.

brentfulgham reviewed (approved):
Thank you for updating the patch. r=me!

vitorroriz committed:
Decoupling from-font resolution from FontCascadeFonts::primaryFont
https://bugs.webkit.org/show_bug.cgi?id=273339
rdar://127132471

Reviewed by Brent Fulgham.

font-size-adjust from-font resolution was fixed by [1]. It now works correctly and
it uses the primary font for resolution. However, this patch changed FontCascadeFonts::primaryFont
to receive a mutable (non-const) FontCascadeDescription object, such that it could resolve the
from-font reference after the primary font is resolved.

This means that now FontCascadeFonts::primaryFont does two different things:
a) resolving and caching a font as the primary font
b) resolving from-font according to the primary font if necessary

The (a) part of primaryFont doesn't require a mutable FontCascadeDescription.
We can update FontCascadeDescription to have a method for resolving font-size-adjust
from-font at the point where we own a mutable FontCascadeDescription instead.

This allows us to compute the primaryFont itself from places where we only have access
to a const FontCascadeDescription. This is for example, required for solving
https://bugs.webkit.org/show_bug.cgi?id=273233, where we need to access to primaryFont
from FontCascadeFonts::glyphDataForVariant

This patch should have no side effects, as it is a preparation
for solving 273233.

[1] https://commits.webkit.org/269041@main

* Source/WebCore/css/ComputedStyleExtractor.cpp:
(WebCore::fontSizeAdjustFromStyle):
* Source/WebCore/platform/graphics/FontCascade.h:
(WebCore::FontCascade::primaryFont const):
* Source/WebCore/platform/graphics/FontCascadeDescription.cpp:
(WebCore::FontCascadeDescription::resolveFontSizeAdjustFromFontIfNeeded):
* Source/WebCore/platform/graphics/FontCascadeDescription.h:
* Source/WebCore/platform/graphics/FontCascadeFonts.h:
(WebCore::FontCascadeFonts::primaryFont):
* Source/WebCore/platform/graphics/FontSizeAdjust.h:
(WebCore::FontSizeAdjust::resolveFromFontIfNeeded):
(WebCore::FontSizeAdjust::shouldResolveFromFont const):

Canonical link: https://commits.webkit.org/278130@main
